### PR TITLE
UCT/DC: Implement random dci allocation policy

### PR DESCRIPTION
**Pull request by brminich**
## What
Add random DCI allocation policy for DC transport.

## Why ?
This allocation policy is known to provide better communication/computation overlap.

## How ?
With this policy every endpoint gets its own DCI during initialization.
Several endpoints may share the DCI.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8632/ for details (Mellanox internal link).

**Comment by brminich** [comment]
@hoopoepg, plz review

File: src/uct/ib/dc/dc_mlx5.c [hunk 1]
```diff
@@ -50,7 +51,10 @@ ucs_config_field_t uct_dc_mlx5_iface_config_sub_table[] = {
      "dcs_quota  Same as \"dcs\" but in addition the DCI is scheduled for release\n"
      "           if it has sent more than quota, and there are endpoints waiting for a DCI.\n"
      "           The dci is released once it completes all outstanding operations.\n"
-     "           This policy ensures that there will be no starvation among endpoints.",
+     "           This policy ensures that there will be no starvation among endpoints.\n"
+     "\n"
+     "rand       Every endpoint is assigned with its own DCI, selected in the random order.\n"
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c]: Every endpoint is assigned with a randomly selected DCI

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 2]
```diff
@@ -968,7 +969,7 @@ ucs_status_t uct_dc_mlx5_ep_pending_add(uct_ep_h tl_ep, uct_pending_req_t *r,
     uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
     uct_dc_mlx5_ep_t *ep = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
 
-    /* ep can tx iff
+    /* ep can tx if
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: why the change?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 3]
```diff
@@ -1061,13 +1065,18 @@ uct_dc_mlx5_iface_dci_do_pending_tx(ucs_arbiter_t *arbiter,
...
+    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+        /* Deschedule the group even if FC is the only resource, which
+         * is missing. It will be scheduled again when credits arrive.
+         * We can't desched group with rand policy if non FC resources are
+         * missing, since it's never scheduled again. */
+        if (uct_dc_mlx5_iface_is_dci_rand(iface) &&
+            uct_rc_fc_has_resources(&iface->super.super, &ep->fc)) {
+            return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+        } else {
+            return UCS_ARBITER_CB_RESULT_DESCHED_GROUP;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: the condition here is changed, even for DCS policy. What is the reason? Maybe should add assertions on  uct_dc_mlx5_ep_fc_wait_for_grant()?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 4]
```diff
@@ -1114,6 +1123,7 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep, uct_pending_purge_callback_t c
     } else {
         ucs_arbiter_group_purge(uct_dc_mlx5_iface_tx_waitq(iface), &ep->arb_group,
                                 uct_dc_mlx5_ep_abriter_purge_cb, &args);
+
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c]: remove?

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 5]
```diff
@@ -231,7 +238,14 @@ static UCS_F_ALWAYS_INLINE ucs_status_t
 uct_dc_mlx5_ep_basic_init(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
 {
     ucs_arbiter_group_init(&ep->arb_group);
-    ep->dci   = UCT_DC_MLX5_EP_NO_DCI;
+
+    if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        /* coverity[dont_call] */
+        ep->dci = rand() % iface->tx.ndci;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h]: must NOT use rand() rand since it changes global rand state and can break applications/tests which check data validity
need to use rand_r()

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 2 above)
Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: typo fix. 
remove?

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 3 above)
Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: not sure I understand the question. For dcs* policies we desched the group always (like it was before).
For rand policy we:
- desched the group of FC resources are missing, because the group will be scheduled again (for any policy) when FC grant arrives
- resched the group if non FC resources are missing, because it will not be scheduled again upon TX completions

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: it's not a typo, iff means "if and only if"

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: is it possible that uct_dc_mlx5_iface_dci_ep_can_send(ep) is false, and uct_dc_mlx5_ep_fc_wait_for_grant(ep) is true?
in this case, before the change, it returned DESCHED, and after, it would return STOP (for dcs*)

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 2 above)
Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: ok, I did not know it

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: i saw you added it in 70df3ea

Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: not sure why we checked for `uct_dc_mlx5_ep_fc_wait_for_grant` originally. It does not meant that FC resources are exhausted, it means that grant request was sent. But `uct_dc_mlx5_iface_dci_ep_can_send(ep)` checks for all ep related resources: QP and FC. Thus:
- if ep can send - it means we don't have iface resources, we have to stop
- if ep can't send we need to deschedule the group (or reschedule in case of FC)

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: i think the idea was that if EP is waiting for grant, it couldn't be flushed, to avoid segfault when getting grant message

Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: but we have list of destroyed ep now. 
I guess this particular change was a mistake in 70df3ea. I should had left check for fc window instead (replaced with inline func which is not exactly the same)

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: ok
